### ISSUE TITLE
systemd gogs-web-1.service not start when reboot (mariadb or mysql is not started yet)

### ISSUE THREAD

**Issue by dariolesca** [gogits/gogs#4223]
- Gogs version (or commit ref):
[root@server-gogs ~]# rpm -q gogs
gogs-0.10.1-1488280184.9d40b8a.centos7.x86_64

- Git version: 
[root@server-gogs ~]# rpm -q git
git-1.8.3.1-6.el7_2.1.x86_64

- Operating system: 
Centos 7.3

- Database (use `[x]`):
  - [ ] PostgreSQL
  - [X] MySQL
  - [ ] SQLite

- Can you reproduce the bug at https://try.gogs.io:
  - [ ] Yes (provide example URL)
  - [X] No
  - [ ] Not relevant

- Log gist (usually found in `log/gogs.log`):
2017/03/02 13:53:15 [TRACE] Log Mode: File (Trace)
2017/03/02 13:53:15 [ INFO] Gogs 0.10.1.0228
2017/03/02 13:53:15 [ INFO] Cache Service Enabled
2017/03/02 13:53:15 [ INFO] Session Service Enabled
2017/03/02 13:53:15 [FATAL] [...s/routers/install.go:67 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: get
sockopt: connection refused
2017/03/02 13:53:15 [TRACE] Log Mode: File (Trace)
2017/03/02 13:53:15 [ INFO] Gogs 0.10.1.0228
2017/03/02 13:53:15 [ INFO] Cache Service Enabled
2017/03/02 13:53:15 [ INFO] Session Service Enabled
2017/03/02 13:53:15 [FATAL] [...s/routers/install.go:67 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: get
sockopt: connection refused
2017/03/02 13:53:16 [TRACE] Log Mode: File (Trace)
2017/03/02 13:53:16 [ INFO] Gogs 0.10.1.0228
2017/03/02 13:53:16 [ INFO] Cache Service Enabled
2017/03/02 13:53:16 [ INFO] Session Service Enabled
2017/03/02 13:53:16 [FATAL] [...s/routers/install.go:67 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: get
sockopt: connection refused
2017/03/02 13:53:16 [TRACE] Log Mode: File (Trace)
2017/03/02 13:53:16 [ INFO] Gogs 0.10.1.0228
2017/03/02 13:53:16 [ INFO] Cache Service Enabled
2017/03/02 13:53:16 [ INFO] Session Service Enabled
2017/03/02 13:53:16 [FATAL] [...s/routers/install.go:67 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: getsockopt: connection refused
2017/03/02 13:53:16 [TRACE] Log Mode: File (Trace)
2017/03/02 13:53:16 [ INFO] Gogs 0.10.1.0228
2017/03/02 13:53:16 [ INFO] Cache Service Enabled
2017/03/02 13:53:16 [ INFO] Session Service Enabled
2017/03/02 13:53:16 [FATAL] [...s/routers/install.go:67 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 127.0.0.1:3306: getsockopt: connection refused

## Description
If I reboot the server, service gogs-web-1.service not start

Services gogs*.service must start after mysql or mariadb is started

I have add and modify 'After=' services entry like this:

[root@server-gogs ~]# grep Aft /etc/systemd/system/gogs.service /etc/systemd/system/gogs-web.service /etc/systemd/system/gogs-web-1.service
/etc/systemd/system/gogs.service:After=mariadb.service
/etc/systemd/system/gogs-web.service:After=mariadb.service gogs.service
/etc/systemd/system/gogs-web-1.service:After=mariadb.service gogs-web.service

and now when I reboot the server gogs start property and work

Dario

**Comment by Unknwon** [comment]
Yes, you did the right thing, but what is the purpose of this issue?

**Comment by dariolesca** [comment]
> Yes, you did the right thing, but what is the purpose of this issue?
My purpose is modify the systemd gogs*.service and add mysql/mariadb.service into After= parameter.
Sorry, I'm not a developer, let me know what I must to do.

**Comment by Unknwon** [comment]
>My purpose is modify the systemd gogs*.service and add mysql/mariadb.service into After= parameter.

Add to where?

**Comment by dariolesca** [comment]
Probably here: https://github.com/gogits/gogs/blob/master/scripts/systemd/gogs.service
you put me in difficulty, I do not are a builder, i'm only a system manager, and I do not able to use git, I do not know if this is the right place where add the "After=" db dependency.
I see that some line have this setting, but this line are commented, and is missing the "After=mariadb.service"
Probably another the right place is the .spec file, used to rebuild the rpm, but I do not found it.
Hope this help to clarify my humble opinion.
Dario

**Comment by Unknwon** [comment]
OK... purpose of those files under `scripts` are stated already: https://github.com/gogits/gogs/tree/master/scripts

But yes, I can another line with `After=mariadb.service` for people to uncomment if they need.

**Comment by Unknwon** [comment]
Please see https://github.com/gogits/gogs/commit/b39454ca16246643530135c519d0437da6090bfc.

**Comment by dariolesca** [comment]
Ok, thanks.
I have post this request to Fedora Devel ML
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/AHKE5HYR24P6UP7A3KG4RSEBZLMMGTC7/

The  Lennart Poettering reply say that is not a problem remove all comment to After= directive.
> After= has no effect if the specified unit doesn#t exist and generates no errors.

Then the my proposal is **remove all comment to After= directive** to avoid modify the gogs.service systemd file

Thanks
Dario

**Comment by Unknwon** [comment]
>Then the my proposal is remove all comment to After= directive to avoid modify the gogs.service systemd file

Is this behavior consistent across all versions of systemd? It sounds very good.

**Comment by dariolesca** [comment]
I have post your question to ML

https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/message/4HGAD4ZZEBA67ERJ52I45ZMUNW2D2MTD/

Tomorrow I try this feature on some my systems and let you know

Thanks
Dario

**Comment by dariolesca** [comment]
I have post this question on Fedora devel ML:

> Then i can remove all comment to After= directive and, if one or more service are not
> installed, systemd do not waits for missing services before start gogs.service.
> Is this behavior consistent across all versions of systemd and all distribution?

The reply from  Lennart Poettering is:

> Yes

([tread here](https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/AHKE5HYR24P6UP7A3KG4RSEBZLMMGTC7/#4HGAD4ZZEBA67ERJ52I45ZMUNW2D2MTD))

Then my proposal is:

Comment out all After= line or put, into [unit] session, all After want service together like this:
`After=mariadb.service mysqld.service postgresql.service memcached.service redis.service`

I have also do a test on my server and all work fine: if I remove the mariadb.service from After= directive and restart the server, gogs-web-1.service do not start, if I add it and reboot the service start and all work fine.

Hope this help.

Dario

**Comment by Unknwon** [comment]
@dariolesca Awesome, would you like propose a PR for this change?

**Comment by dariolesca** [comment]
Yes, but I'm not a git expert.
What does it mean PR?
Give me some suggest and I do it, or propose you a PR like two day ago (Please see b39454c.)
Let me know
Thanks ... Unknwon
Dario

**Comment by Unknwon** [comment]
Ah... sorry, never mind I'll do it for you!

**Comment by Unknwon** [comment]
Please see https://github.com/gogits/gogs/commit/7e883f891a42fd01bc54ce1821d3520d691f1a63.

**Comment by dariolesca** [comment]
OK, I have see.
Many Thanks for support, Unknwon !
Have a nice Day!!

**Comment by raabf** [comment]
I can verify that the new `gogs.service` is working now (I did have also a mysql dependency problem on start)

However your package script seems to not include the file into packager.io (at least for the debian packages, tested with the 11.4 deb package)

instead various probably auto generated files are present in /etc/systemd/system and used by the system:

   ├── gogs.service
   ├── gogs.service.wants
   │   └── gogs-web.service -> /etc/systemd/system/gogs-web.service
   ├── gogs-web-1.service
   ├── gogs-web.service
   ├── gogs-web.service.wants
   │   └── gogs-web-1.service -> /etc/systemd/system/gogs-web-1.service

basically all of them do nothing really, except that gogs-web requires gogs. For example `gogs.service` looks as follows:

    [Unit]
    StopWhenUnneeded=true

    [Service]
    # this service is just a placeholder
    ExecStart=/bin/sleep infinity
    
    [Install]
    WantedBy=multi-user.target

Especially the `After=` are still missing and in my case mysql is loaded after gogs and gogs will fail to start.
I can add the `After=` After lines as proposed in 7e883f8 to this file and it works now correctly during start.

**Comment by JohnMyue** [comment]
hi,i met this issue ,i use docker gogs,i can't find gogs.service, i need help.
`2018/07/16 23:43:02 [ INFO] Cache Service Enabled
2018/07/16 23:43:02 [ INFO] Session Service Enabled
2018/07/16 23:45:09 [FATAL] [...gs/routes/install.go:66 GlobalInit()] Fail to initialize ORM engine: migrate: sync: dial tcp 180.76.57.134:3306: getsockopt: connection timed out
2018/07/16 23:45:09 [TRACE] Log Mode: File (Trace)`

**Comment by Unknwon** [comment]
@JohnMyue 

>dial tcp 180.76.57.134:3306: getsockopt: connection timed out
